pkg/broadcaster/unbuffered.go

```diff
@@ -1,49 +1,141 @@
 func (w *Unbuffered) Add(writer io.WriteCloser) {
-	w.mu.Lock()
-	w.writers = append(w.writers, writer)
-	w.mu.Unlock()
+	w.writersWriteMu.Lock()
+	w.writersUpdateMu.Lock()
+
+	if w.writers == nil {
+		w.writers = new([]io.WriteCloser)
+	}
+	*w.writers = append(*w.writers, writer)
+
+	w.writersUpdateMu.Unlock()
+	w.writersWriteMu.Unlock()
 }
 
 // Write writes bytes to all writers. Failed writers will be evicted during
 // this call.
 func (w *Unbuffered) Write(p []byte) (n int, err error) {
-	w.mu.Lock()
+	w.writersWriteMu.Lock()
+	defer w.writersWriteMu.Unlock()
+
+	w.writersUpdateMu.Lock()
+
+	// make a copy of w.writers. Even if .CleanContext sets w.writers, this
+	// Write call will still use it's valid copy
+	writers := w.writers
+
+	// release w.writersUpdateMu. This allows CleanContext to set w.writers
+	// to nil - but this Write call won't be affected by this, as it uses a copy
+	// of that pointer. We will also be able to safely iterate over *writers:
+	// clean methods never resize the slice (they only may set pointer to nil),
+	// and Add and Write require w.writersWriteMu, which we are still holding
+	w.writersUpdateMu.Unlock()
+
+	if writers == nil {
+		return
+	}
+
 	var evict []int
-	for i, sw := range w.writers {
+	for i, sw := range *writers {
 		if n, err := sw.Write(p); err != nil || n != len(p) {
 			// On error, evict the writer
 			evict = append(evict, i)
 		}
 	}
+
+	w.writersUpdateMu.Lock()
+	// at this point w.writers might have already been set to nil, but we're
+	// not affected by this as we are using a copy
 	for n, i := range evict {
-		w.writers = append(w.writers[:i-n], w.writers[i-n+1:]...)
+		*writers = append((*writers)[:i-n], (*writers)[i-n+1:]...)
 	}
-	w.mu.Unlock()
+	w.writersUpdateMu.Unlock()
+
 	return len(p), nil
 }
 
+func (w *Unbuffered) cleanUnlocked() {
+	if w.writers == nil {
+		return
+	}
+	for _, sw := range *w.writers {
+		sw.Close()
+	}
+	w.writers = nil
+}
+
+func (w *Unbuffered) cleanWithWriteLock() {
+	w.writersUpdateMu.Lock()
+	w.cleanUnlocked()
+	w.writersUpdateMu.Unlock()
+}
+
 // Clean closes and removes all writers. Last non-eol-terminated part of data
 // will be saved.
 func (w *Unbuffered) Clean() error {
-	w.mu.Lock()
-	for _, sw := range w.writers {
-		sw.Close()
+	w.writersWriteMu.Lock()
+
+	w.cleanWithWriteLock()
+
+	w.writersWriteMu.Unlock()
+	return nil
+}
+
+// CleanContext closes and removes all writers.
+// CleanContext supports timeouts via the context to unblock and forcefully
+// close the io streams. This function should only be used if all writers
+// added to Unbuffered support concurrent calls to Close and Write: it will
+// call Close while Write may be in progress in order to forcefully close
+// writers
+func (w *Unbuffered) CleanContext(ctx context.Context) error {
+	// there is essentially a race between normal cleaning (done while holding
+	// w.writersWriteMu) and forceful one - cleaningUnderway is used to tell
+	// who won the race and avoid double cleanup
```

Comment:
The implementation could be greatly simplified, eliminating the race, [if `w.writersWriteMu` was a 1-buffered `chan struct{}` instead of a classical mutex.](https://drive.google.com/file/d/1nPdvhB0PutEJzdCq5ms6UI58dp50fcAN/view)
```go
select {
case <-w.writersWriteMu:
	defer func() { w.writersWriteMu <- struct{}{} }()
case <-ctx.Done():
}
w.cleanWithWriteLock()
return ctx.Err()
```

Reply:
Thanks for the idea! -)

I'll look into this

Reply:
@corhere I have tried this idea. `CleanContext` indeed looks much better after this change. However, there is an issue here - unlike a mutex, a channel requires initialization: it should be created with `make()`. `broadcaster.Unbuffered` does not currently have a constructor, so we have 2 options here:
* create `w.writersWriteMu` channel on demand. This will create a new race condition - a race for the `w.writersWriteMu` field itself. For example, if two goroutines call `.Add()` on a fresh `Unbuffered` instance at the same time, there will be a race in assignment of the created `w.writersWriteMu`. So, we will need an additional lock that protects this field. An implementation with this additional lock can look like this https://gist.github.com/asterite3/eb938f97f58a02a14916cebfaa174373 (this is not well tested yet, but I think it would more-or-less like this)
* add a constructor function `NewUnbuffered` to `broadcaster` package. Then, all users of this package should be updated to start using this function. I am not entirely sure how hard is that. On the one hand, in moby repository it seems that there is only one place where `Unbuffered` is created: `container/stream/streams.go`. On the other hand, this is a public package and other users may exist (although by looking at [godoc](https://pkg.go.dev/github.com/docker/docker/pkg/broadcaster?tab=importedby) it seems only forks of moby are using it). Are there any guarantees about the stability of packages in `pkg/`? If adding a (mandatory) constructor function is OK, it seems a better option.

@laurazard what do you think ?

Reply:
You're right, `sync.Once` can be used instead of an additional lock (`writersWriteMuMu` in my implementation). And on-demand initialization will look much simpler then.

Reply:
Is this an implementation variant we want to proceed with?
I have switched to it in this PR.

(The previuos implementation that does not require an extra `sync.Once` but where `CleanContext` is complicated and uses atomics is [left here](https://gist.github.com/asterite3/069b11ee1d41ecbe9e926593c1aa30b1) just in case)

Reply:
Btw, `sync.Once` can also be used to simplify `CleanContext` in variant where `writersWriteMu` is a regular `sync.Mutex`: it can [look like this](https://gist.github.com/asterite3/44a3c6ab93b43888e0ba8c2142ce5d02#file-unbuffered-go-L105). Also, as far as I understand, after this change `CleanContext` is not just simpler, the semantics are changed so that `CleanContext` never returns until `w.cleanWithWriteLock` is done - which is a right thing I guess. And in my initial implementation `CleanContext` could return before `w.cleanWithWriteLock` returned (i. e. before cleanup is done), which seems to be a bug.

Reply:
Sorry, have been out on leave for a while will take a look. Generally, re:
> Is this an implementation variant we want to proceed with?

I think the answer is yes, since it seems both cleaner and more correct. I'll try to review it soon :)

Reply:
> Sorry, have been out on leave for a while will take a look. Generally, re:
> 
> > Is this an implementation variant we want to proceed with?
> 
> I think the answer is yes, since it seems both cleaner and more correct. I'll try to review it soon :)

Okay! -)

The variant with atomics is certainly worse and should not be chosen, but I encourage you to also check out the variant linked from my previous comment, which also uses `sync.Once`, but inside `CleanContext`, and where `writersWriteMu` is a `sync.Mutex`